atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/priority/NextTableToSweepProviderImpl.java

```diff
@@ -63,47 +83,126 @@ public NextTableToSweepProviderImpl(KeyValueService kvs, SweepPriorityStore swee
+                double priority = getSweepPriority(oldPriority, newPriority);
+                tableToPriority.put(oldPriority.tableRef(), priority);
+            } else {
+                toDelete.add(oldPriority.tableRef());
             }
+        }
+
+        // Clean up rows for tables that no longer exist.
+        sweepPriorityStore.delete(tx, toDelete);
+
+        logPrioritiesByTable(tableToPriority);
+
+        List<TableReference> tablesWithHighestPriority = findTablesWithHighestPriority(tableToPriority);
+        return tablesWithHighestPriority.size() > 0 ?
+                Optional.of(getRandomValueFromList(tablesWithHighestPriority)) :
+                Optional.empty();
+    }
 
-            // Clean up rows for tables that no longer exist.
-            sweepPriorityStore.delete(tx, toDelete);
-            return toSweep;
+    private void logUnsweptTables(Set<TableReference> unsweptTables) {
+        if (!log.isDebugEnabled()) {
+            return;
         }
+
+        LoggingArgs.SafeAndUnsafeTableReferences safeAndUnsafeTableReferences = LoggingArgs.tableRefs(unsweptTables);
+
+        log.debug("Unswept tables: {} and {}",
+                SafeArg.of("tables", safeAndUnsafeTableReferences.safeTableRefs()),
+                SafeArg.of("unsafeTables", safeAndUnsafeTableReferences.unsafeTableRefs()));
+    }
+
+    private void logPrioritiesByTable(Map<TableReference, Double> tableToPriority) {
+        if (!log.isDebugEnabled()) {
+            return;
+        }
+
+        List<Map.Entry<TableReference, Double>> tablesSortedByPriority = tableToPriority.entrySet().stream()
+                .sorted(Comparator.comparingDouble(Map.Entry::getValue))
+                .collect(Collectors.toList());
+
+        List<Map.Entry<TableReference, Double>> safeTablesEntries = new ArrayList<>();
+        List<Map.Entry<TableReference, Double>> unsafeTablesEntries = new ArrayList<>();
+
+        for (Map.Entry<TableReference, Double> tableEntry : tablesSortedByPriority) {
+            if (LoggingArgs.tableRef(tableEntry.getKey()).isSafeForLogging()) {
+                safeTablesEntries.add(tableEntry);
+            } else {
+                unsafeTablesEntries.add(tableEntry);
+            }
+        }
+
+        log.debug("Sweep priorities per table: {} and {}",
+                SafeArg.of("tables", safeTablesEntries.toString()),
+                SafeArg.of("unsafeTables", unsafeTablesEntries.toString()));
+    }
+
+    // TODO(ssouza): move this to the BackgroundSweeperImpl when we make this method return a list.
+    private List<TableReference> findTablesWithHighestPriority(
+            Map<TableReference, Double> tableToPriority) {
+        Double maxPriority = Collections.max(tableToPriority.values());
+        return tableToPriority.entrySet().stream()
+                .filter(entry -> Objects.equals(entry.getValue(), maxPriority))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
```

Comment:
Not sure if this code is a little to much functional... Started writing it in the old way, but changed to this as the final piece. I see @jeremyk-91's functional programming influencing me. Maybe a bit too much?